typing_extensions/src_py3/typing_extensions.py

```diff
@@ -985,6 +997,188 @@ def __new__(cls, *args, **kwds):
         Protocol.__doc__ = Protocol.__doc__.format(bases="Protocol, Generic[T]" if
                                                    OLD_GENERICS else "Protocol[T]")
 
+
+if PEP_560:
+    # We need to rewrite these from scratch.
+    del _ProtocolMeta, Protocol
+
+    from typing import _type_check, _GenericAlias, _collect_type_vars
+
+    class _ProtocolMeta(abc.ABCMeta):
+        # This metaclass is a bit unfortunate and exists only because of the lack
+        # of __instancehook__.
+        def __instancecheck__(cls, instance):
+            # We need this method for situations where attributes are
+            # assigned in __init__.
+            if ((not getattr(cls, '_is_protocol', False) or
+                    _callable_members_only(cls)) and
+                    issubclass(instance.__class__, cls)):
+                return True
+            if cls._is_protocol:
+                if all(hasattr(instance, attr) and
+                        (not callable(getattr(cls, attr, None)) or
+                         getattr(instance, attr) is not None)
+                        for attr in _get_protocol_attrs(cls)):
+                    return True
+            return super().__instancecheck__(instance)
+
+
+    class Protocol(metaclass=_ProtocolMeta):
+        # There is quite a few overlapping code with typing.Generic.
+        # Unfortunarely it is hard to avoid this while these live in two different modules.
+        # The duplicated code will be removed when Protocol is moved to typing.
```

Comment:
We can move it already in 3.7.1, however this will require `_ProtocolMeta`, since we can add `__instancehook__` only in 3.8. Although this is probably not a big deal, since I expect that explicit subclassing of protocols by concrete classes will be rare. One of the main goals of PEP 544 is to allow not to subclass them :-)